Feature: Rocket Cognito Auth (#556)
boostercloud/booster: packages/rocket-auth-aws-infrastructure/README.md

# Authentication Booster Rocket for AWS

> This package is a configurable Booster rocket to add an authentication api based on Cognito to your Booster applications.

Reviewer: ```suggestion
This package is a configurable Booster rocket to add an authentication API based on Cognito to your Booster applications.
```

## Usage

> Install this package as a dev dependency in your Booster project (It's a dev dependency because it's only used during deployment, but we don't want this code to be uploaded to the project lambdas)

Reviewer: ```suggestion
Install this package as a `devDependency` in your Booster project (It's a `devDependency` because it's only used during deployment, so we don't want this code to be uploaded to the project lambdas)
```

```sh
yarn add --dev @boostercloud/rocket-auth-aws-infrastructure
```

In your Booster config file, pass a `RocketDescriptor` array to the `AWS Provider` initializer configuring the aws authentication rocket:

```typescript
import { Booster } from '@boostercloud/framework-core'
import { BoosterConfig } from '@boostercloud/framework-types'
import * as AWS from '@boostercloud/framework-provider-aws'

Booster.configure('production', (config: BoosterConfig): void => {
  config.appName = 'my-store'
  config.provider = Provider([
    {
      packageName: '@boostercloud/rocket-auth-aws-infrastructure',
      parameters: {
        config,                  
        mode: 'Passwordless',                     
      },
    },
  ])
})
```

## Configuration Options

```typescript
{
  config: BoosterConfig                      // Required
  passwordPolicy?: {                         // Optional, all values are set to true by default.
    minLength?: number                       // Minimum length, which must be at least 6 characters but fewer than 99 character
    requireDigits: boolean                   // Require numbers
    requireLowercase: boolean                // Require lowercase letters

>     requireSymbols: boolean                  // Require a special characters

Reviewer: ```suggestion
    requireSymbols: boolean                  // Require special characters
```

    requireUppercase: boolean                // Require uppercase letters
  }

>   mode: 'Passwordless' | 'UserPassword'      // One time password the user must be a phone number | User Password mode, the user must be a valid email.

Reviewer: ```suggestion
  mode: 'Passwordless' | 'UserPassword'      // If Passwordless, the user must be a phone number. If UserPassword, the user must be a valid email.
```

#### Endpoint

```http request
POST https://<httpURL>/auth/sign-up/confirm
```

#### Request body

```json
{
  "confirmationCode": "string",
  "username": "string"
}
```

| Parameter          | Description                                                                            |
| ------------------ | -------------------------------------------------------------------------------------- |
| _confirmationCode_ | The confirmation code received in the SMS message.                                     |
| _username_         | The username of the user you want to sign in. They must have previously signed up.     |

#### Response

```json
{
  "message": "The username: +34999999999 has been confirmed."
}
```

#### Errors

You will get an HTTP status code different from 2XX and a body with a message telling you the reason of the error.
Common errors would be like submitting an expired confirmation code or a non valid one.

Example of an invalid verification code:

```json
{
  "error": {
    "type": "CodeMismatchException",
    "message": "Invalid verification code provided, please try again."
  }
}
```

### Resend Sign-up confirmation code

> If for some reason the confirmationCode never reach your email or your phone via SMS you can ask the api to resend a new one.

Reviewer: ```suggestion
If for some reason the confirmation code never reaches your email, or your phone via SMS, you can ask the API to resend a new one.
```
